### ISSUE TITLE
Provisioner remote-exec Terraform 

### ISSUE THREAD

**Issue by Karthick-Gopalakrishnan** [QT-DevOps/DevOpsIssues#472]
Hello Team,

Am trying to execute the below provisioner in Azure where am not getting the public ip 

provisioner "remote-exec" {
  inline = ["echo myip [${azurerm_public_ip.mypublicip.ip_address}] >> /home/azureuser/myinvent"]
}

It wanted to create a file by the myinvent with content myip [xx.xx.xx.xx] . above provisioner is creating a file by the name myinvent with content myip [ ] Public ip seems to be missing.

did anyone find a way around to print the public ip ?! 

P.S : There was a wait time of 2 Min before executing the provisioner still the ip block was empty

**Comment by bhaskardegala** [comment]
If you have selected dynamic in public ip address allocation then you need to link your public ip to your VM and even you need to re-perform terraform apply again to see your public ip address 
To make things simple select static public ip address allocation then ip address reflect there even though you don't have an vm

**Comment by Karthick-Gopalakrishnan** [comment]
Bhaskar, 

I also have an output block , When i reapply terrafrom i do get the IP address in output block, but wat am trying to do here is to create a file with Public ip inside the VM

output "webserverip" {
  value = "${azurerm_public_ip.mypublicip.ip_address}"
}

So your point is if i configure the terrafrom publicip block to Static , I will get the ip address printed in a file inside the vm ?

**Comment by bhaskardegala** [comment]
Then try using variables ie
variables "ip_address" {
    value = "${azurerm_public_ip.mypublicip.ip_address}"
}
And use this variable in your command 
Example :- echo ${var.ip_address}
If not try traditional way to find ip address in Linux you can refer this [link](https://opensource.com/article/18/5/how-find-ip-address-linux)  for that

**Comment by ghost** [comment]
if you want to print public IP in remote-exec  provisioner section  use shell scripting 
remote-exec dont know about terraform DSL

**Comment by Karthick-Gopalakrishnan** [comment]
Thanks for the input Naresh & Bhaskar,

I was trying to print the Public ip inside the ansible inventory file , i will work on ways you suggested. 

Did you guys try to build centos in Azure using Terraform ?!

when trying to push the yum update and other yum related command in remote-exec it is going to loop asking for sudo password same while executing the shell script , where you able to by pass it ?

**Comment by bhaskardegala** [comment]
> Thanks for the input Naresh & Bhaskar,
> 
> I was trying to print the Public ip inside the ansible inventory file , i will work on ways you suggested.
> 
> Did you guys try to build centos in Azure using Terraform ?!
> 
> when trying to push the yum update and other yum related command in remote-exec it is going to loop asking for sudo password same while executing the shell script , where you able to by pass it ?

While you are trying to use ansible in terraform there is no need to write anything inside the inventory file  , in your Playbook just enter 
hosts: localhost 
and if you are trying use ip inside your playbook then use this "ansible_default_ipv4.address"and I will look into yum issues

**Comment by Karthick-Gopalakrishnan** [comment]
Bhaskar,

Is used loopback address IP in inventory file and Host: all, This setup worked. 

Sent from Mail for Windows 10

**Comment by sandhyadev** [comment]
After "terraform apply" completed successfully, if there are no warnings or errors and ur resources are deployed exactly as expected. The output does not show up at all. Only after u run 'terraform refresh', the output file with ip address is  displayed.

**Comment by Karthick-Gopalakrishnan** [comment]
I saw a blogs where they were invoking the publicip inside local-exec so i tried with remote-exec  and i didnt do it right 

![image](https://user-images.githubusercontent.com/48745218/58367921-44a17f80-7ef6-11e9-8f99-ebc85d99bdb7.png)

provisioner "remote-exec" {
inline = ["echo myip [${azurerm_public_ip.mypublicip.ip_address}] >> /home/azureuser/myinvent"]
}

Thanks for all your reply !!, I have completed what i was looking for.. Keeping it open Since i had another issue with yum install in centos VM if you have tried deploying centos VM with yum upgrade or running script without password loop, Kindly let me know .

**Comment by bhaskardegala** [comment]
Try using this 
provisioner "remote-exec" {
    inline = [
      "sudo amazon-linux-extras enable nginx1.12",
      "sudo yum -y install nginx",
      "sudo systemctl start nginx",
    ]
  }

**Comment by Karthick-Gopalakrishnan** [comment]
Hello Baskar,

I tried with below Provisioner in AZURE (not AWS) 
provisioner "remote-exec" {
  inline = [ "sudo yum install tree -y",
            "sudo yum upgrade -y "
            ] }

as said before it was going in loop asking for SUDO password. 

I got below findings the user id created is not in the wheel or root groups. I tried to provision using ROOT as user id it was stopping with a  Message="The Admin Username specified is not allowed.

PFA images for your reference.

[Centos.zip](https://github.com/QT-DevOps/DevOpsIssues/files/3220035/Centos.zip)

**Comment by bhaskardegala** [comment]
even i was facing this issue with yum gopal i will let u know once i find out resolution

**Comment by bhaskardegala** [comment]
hey gopal, 
i tried something and it can definitely help you as it helped me
 ---

  connection {
    type     = "ssh"
    user     = "${var.username}"
    password = "${var.password}"
  }
  provisioner "remote-exec" {
    inline = [
      "echo ${var.password} | sudo -S yum upgrade -y",
      "echo ${var.password} | sudo -S yum install tree -y"
    ]
  }

---

**Comment by Karthick-Gopalakrishnan** [comment]
Bhaskar,

It worked, Thanks for your inputs !!

**Comment by Karthick-Gopalakrishnan** [comment]
Thanks Naresh & Sandhya, I will close the issue
